### DISCUSSION TITLE
feature request: pdk new function

### DISCUSSION THREAD

**Discussion by scotje** [puppetlabs/pdk#1023] [Feature Request]
_From @andrewm3 on March 8, 2018 22:25_

While support for Puppet DSL items are quite good, it would be nice to have some Ruby templates e.g. the ability to create a new Ruby function using the modern API.

It would also be great to stub the rspec boilerplate for the function.

_Copied from original issue: puppetlabs/pdk#450_

**Comment by scotje** [comment]
_From @rodjek on April 4, 2018 23:21_

Being tracked in [JIRA](https://tickets.puppetlabs.com/browse/PDK-700)

**Comment by logicminds** [comment]
I have been using my own generator for years.  I would like to see the same thing in pdk.  The only tricky bit is you need to parse the folder tree from the given name and also enforce a namespace.  I would assume we can drop the v3.  

Example template is here: https://github.com/nwops/retrospec-templates/tree/master/functions, although retrospec produces a context to pass into this template. 

```
retrospec puppet new_function --help
Generates a new function with the given name.
  -n, --name=<s>           The name of the function you wish to create
  -t, --type=<s>           The version type of the function (v3,v4,native) (default: v4)
  -u, --test-type=<s>      The type of test file to create (rspec, ruby) (default: rspec)
  -r, --return-type=<s>    The return type of the function (rvalue, statement) (default: rvalue)
  -h, --help               Show this message
```
